src/librbd/AioRequest.cc

```diff
@@ -362,32 +406,111 @@ namespace librbd {
   }
 
   int AbstractWrite::send() {
-    ldout(m_ictx->cct, 20) << "send " << this << " " << m_oid << " " << m_object_off << "~" << m_object_len << dendl;
+    ldout(m_ictx->cct, 20) << "send " << this << " " << m_oid << " "
+			   << m_object_off << "~" << m_object_len << dendl;
+
+    if (send_pre()) {
+      return 0;
+    } else {
+      send_write();
+    }
+    return 0;
+  }
+
+  bool AbstractWrite::send_pre() {
+    bool lost_exclusive_lock = false;
+    {
+      RWLock::RLocker l(m_ictx->owner_lock);
+      RWLock::RLocker l2(m_ictx->md_lock);
+      if (m_ictx->object_map == NULL) {
+	return false;
+      }
+
+      if (!m_ictx->image_watcher->is_lock_owner()) {
+	ldout(m_ictx->cct, 1) << "lost exclusive lock during write" << dendl;
+	lost_exclusive_lock = true;
+      } else {
+	ldout(m_ictx->cct, 20) << "send_pre " << this << " " << m_oid << " "
+			       << m_object_off << "~" << m_object_len << dendl;
+
+        uint8_t new_state;
+        boost::optional<uint8_t> current_state;
+        pre_object_map_update(&new_state);
+
+        m_state = LIBRBD_AIO_WRITE_PRE; 
+        FunctionContext *ctx = new FunctionContext(
+          boost::bind(&AioRequest::complete, this, _1));
+        m_ictx->object_map->aio_update(m_object_no, new_state,
+				       current_state, ctx);
+      }
+    }
+    
+    if (lost_exclusive_lock) {
+      complete(-ERESTART);
+    }
+    return true;
+  }
+
+  bool AbstractWrite::send_post() {
+    ldout(m_ictx->cct, 20) << "send_post " << this << " " << m_oid << " "
+			   << m_object_off << "~" << m_object_len << dendl;
+
+    RWLock::RLocker l(m_ictx->owner_lock);
+    RWLock::RLocker l2(m_ictx->md_lock);
+    if (m_ictx->object_map == NULL || !post_object_map_update()) {
+      return true;
+    }
+
+    if (m_ictx->image_watcher->is_lock_supported() &&
+        !m_ictx->image_watcher->is_lock_owner()) {
+      // leave the object flagged as pending
+      ldout(m_ictx->cct, 1) << "lost exclusive lock during write" << dendl;
+      return true;
+    }
+
+    m_state = LIBRBD_AIO_WRITE_POST;
+    FunctionContext *ctx = new FunctionContext(
+      boost::bind(&AioRequest::complete, this, _1));
+    m_ictx->object_map->aio_update(m_object_no, OBJECT_NONEXISTENT,
```

Comment:
thinking again, we can fix this case later (full object removal isn't normally done, even with discard enabled, since the discard requests are usually split up into smaller chunks by the block layer in a guest).

I'm thinking we'd want a separate PENDING_REMOVAL state, so that a discard followed by a write can't assert the same PENDING state and end up with an inconsistent final state.

This isn't a big deal for trimming since that's not safe to do concurrently with writes anyway.